src/include/miopen/graphapi/engine.hpp

```diff
@@ -27,16 +27,98 @@
 #pragma once
 
 #include <miopen/graphapi/graphapi.hpp>
+#include <miopen/graphapi/tensor.hpp>
+#include <miopen/graphapi/variant_pack.hpp>
 #include <miopen/solution.hpp>
 
+#include <memory>
+#include <string_view>
+
 namespace miopen {
 
 namespace graphapi {
 
+class Engine;
+class OpGraph;
+
+// Pattern is a family of solvers for the same graph shape
+class GraphPatternMatcher
+{
+
+public:
+    virtual bool matches(const OpGraph* graph) const             = 0;
+    virtual std::vector<Engine> getEngines(OpGraph* graph) const = 0;
+    virtual std::string_view name() const                        = 0;
+
+    virtual ~GraphPatternMatcher();
+};
+
+struct TensorInfo
+{
+    miopenTensorArgumentId_t mEnumId = miopenTensorArgumentIdInvalid;
+    Tensor* mGraphTensor             = nullptr;
+    TensorDescriptor mTensDesc{};
+    Data_t mDevBuf = nullptr;
+
+    TensorInfo(miopenTensorArgumentId_t enum_id, Tensor* tens_ptr)
+        : mEnumId(enum_id),
+          mGraphTensor(tens_ptr),
+          mTensDesc(static_cast<TensorDescriptor>(*tens_ptr))
+    {
+        assert(tens_ptr);
+        assert(mEnumId != miopenTensorArgumentIdInvalid);
+    }
+
+    void setDevBuf(Data_t ptr)
+    {
+        assert(ptr);
+        mDevBuf = ptr;
+    }
+
+    const TensorDescriptor* tensDescPtr() const { return &mTensDesc; }
+};
+
+// int64_t is the graph tensor id
+using TensorInfoMap = std::unordered_map<int64_t, TensorInfo>;
+
+class GraphPatternExecutor
+{
+
+public:
+    virtual void execute(miopenHandle_t handle, const VariantPack& vpk) = 0;
+    virtual size_t getWorkspaceSize() const                             = 0;
+    virtual ~GraphPatternExecutor();
+};
+
+// generic executor that uses Find 2.0 Solution
+class GraphExecutorFind20 : public GraphPatternExecutor
+{
+    miopenSolution_t mSolution;
+    std::shared_ptr<TensorInfoMap> mTensorInfoMap;
+
+public:
+    GraphExecutorFind20(miopenSolution_t sol, const std::shared_ptr<TensorInfoMap>& tmap)
+        : GraphPatternExecutor(), mSolution(sol), mTensorInfoMap(tmap)
+    {
+    }
+
+    void execute(miopenHandle_t handle, const VariantPack& vpk) final;
+
+    size_t getWorkspaceSize() const final;
+
+    static std::unique_ptr<GraphPatternExecutor> make(miopenSolution_t sol,
+                                                      const std::shared_ptr<TensorInfoMap>& tmap)
+    {
+        GraphPatternExecutor* p = new GraphExecutorFind20(sol, tmap);
+        return std::unique_ptr<GraphPatternExecutor>(p);
+    }
+};
+
 class Engine
 {
 private:
-    Solution mSolution;
+    std::shared_ptr<GraphPatternExecutor> mExecutor;
+    OpGraph* mGraph;
```

Comment:
```suggestion
    OpGraph* mGraph      = nullptr;
```
Fix uninitialized member